File: 146-lru-cache/lru-cache.cs
using System;
using System.Collections.Generic;

public class LRUCache
{
    private readonly int capacity;
    private readonly Dictionary<int, LinkedListNode<(int key, int value)>> cache;
    private readonly LinkedList<(int key, int value)> usageOrder;

    public LRUCache(int capacity)
    {
        this.capacity = capacity;
        cache = new Dictionary<int, LinkedListNode<(int key, int value)>>();
        usageOrder = new LinkedList<(int key, int value)>();
    }

    public int Get(int key)
    {
        if (!cache.ContainsKey(key))
        {
            return -1;
        }

        // Move the accessed item to the front of the usage order
        var node = cache[key];
        usageOrder.Remove(node);
        usageOrder.AddFirst(node);

        return node.Value.value;
    }

    public void Put(int key, int value)
    {
        if (cache.ContainsKey(key))
        {
            // Update the existing item's value and move it to the front of the usage order
            var node = cache[key];
            node.Value = (key, value);
            usageOrder.Remove(node);
            usageOrder.AddFirst(node);
        }
        else
        {
            if (cache.Count >= capacity)
            {
                // Remove the least recently used item from the cache and usage order
                var lruNode = usageOrder.Last;
                cache.Remove(lruNode.Value.key);
                usageOrder.RemoveLast();
            }

            // Add the new item to the cache and usage order
            var newNode = new LinkedListNode<(int key, int value)>((key, value));
            cache[key] = newNode;
            usageOrder.AddFirst(newNode);
        }
    }
}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.Get(key);
 * obj.Put(key,value);
 */

File: 146-lru-cache/lru-cache.js
class Node{
    constructor(key, value){
        this.key=key;
        this.value=value;
        this.prev=null;
        this.next=null;
    }
}
/**
 * @param {number} capacity
 */
var LRUCache = function(capacity) {
    this.capacity=capacity;
    this.cache=new Map();
    this.head=new Node(null,null);
    this.tail=new Node(null, null);
    this.head.next=this.tail;
    this.tail.prev=this.head;
};

/** 
 * @param {number} key
 * @return {number}
 */
 LRUCache.prototype._remove=function(node){
    node.prev.next=node.next;
    node.next.prev=node.prev;
 };
 LRUCache.prototype._insert=function(node){
    node.next=this.head.next;
    node.prev=this.head;
    this.head.next.prev=node;
    this.head.next=node;
 };
LRUCache.prototype.get = function(key) {
    if(!this.cache.has(key)) return -1;
    const node=this.cache.get(key);
    this._remove(node);
    this._insert(node);
    return node.value;
};

/** 
 * @param {number} key 
 * @param {number} value
 * @return {void}
 */
LRUCache.prototype.put = function(key, value) {
    if(this.cache.has(key)){
        this._remove(this.cache.get(key));
    }
    const newNode=new Node(key,value);
    this._insert(newNode);
    this.cache.set(key,newNode);
    if(this.cache.size>this.capacity){
        const lru=this.tail.prev;
        this._remove(lru);
        this.cache.delete(lru.key);
    }
};

/** 
 * Your LRUCache object will be instantiated and called as such:
 * var obj = new LRUCache(capacity)
 * var param_1 = obj.get(key)
 * obj.put(key,value)
 */